### PR TITLE
wren-console 0.2.90 (new formula)

### PR DESCRIPTION
**Pull request by joshgoebel**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

**Comment by carlocab** [comment]
```
wren-console:
  * Don't use 'git' as a dependency (it's always available)
  * GitHub fork (not canonical repository)
Error: 2 problems in 1 formula detected
```
We don't need the `depends_on "git"`, and we definitely shouldn't be doing `git clone` in the `install` method, as we want to use the versioned and checksummed tarball declared in the `url`.

Doing `git clone` means builds are not reproducible, and skips checksum verification.

**Comment by nandahkrishna** [comment]
Also, this would still fail the notability audit even if it weren't a fork as it does not have 75+ stars or 30+ forks/watchers (the 13 forks shown are because the repository itself is a fork).

File: Formula/wren-console.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+  def install
+    on_macos do
+      system "git", "clone", "https://github.com/joshgoebel/wren-essentials", "deps/wren-essentials"
+      system "make", "-C", "projects/make.mac"
+    end
+    on_linux do
+      system "git", "clone", "https://github.com/joshgoebel/wren-essentials", "deps/wren-essentials"
+      system "make", "-C", "projects/make"
+    end
```
Comment (carlocab) [inline Formula/wren-console.rb:18]:
```suggestion
    project_dir = "projects/make.mac"
    on_linux { project_dir = "projects/make" }
    system "make", "-C", project_dir
```
Something like this should work.

**Review by carlocab — COMMENTED** [review]
Thanks for your PR, @joshgoebel.

This seems like a nice piece of software, but it currently doesn't meet the requirements for inclusion into Homebrew/core. (See Nanda's comment about our notability requirement.)

You are welcome to submit this for inclusion again when it does. When you do so, please address my feedback regarding the `install` method, and regarding the use of Homebrew `wren` and `libuv`.

If you wish, you can host this formula in your own tap in the meantime. Here are some docs to help you get started:

https://docs.brew.sh/Taps
https://docs.brew.sh/Interesting-Taps-and-Forks
https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap
https://brew.sh/2020/11/18/homebrew-tap-with-bottles-uploaded-to-github-releases/

**Comment by joshgoebel** [comment]
> We don't need the depends_on "git", and we definitely shouldn't be doing git clone in the install method, as we want to use the versioned and checksummed tarball declared in the url.

Ok, that's understandable.   Yet `essentials` is a build dependency of `wren-console` but it currently makes no sense at all to package it on its own.  I can tag a versioned release of `essentials` and pull a tarball, but does Homebrew have a way to track the URL and SHA for the second tarball?

> Also, I see that this bundles wren and libuv. We should use Homebrew's versions of wren and libuv here instead.

_The [official `wren-cli` package](https://formulae.brew.sh/formula/wren-cli   ) does this exact same thing._ 

If `wren-cli` can bundle critical dependencies then surely we could also?  The console tools are built/linked (and tested) against _specific_ versions of `wren` and `libuv` on purpose - they aren't intended to be upgraded constantly.  Reproducible builds and all. :-)

- `wren-console` also uses a custom patched version of Wren 0.4 to support needed functionality.

> See Nanda's comment about our notability requirement.)

Understood.  I think that's solvable, so I'm more concerned with the other matters.  Plus I assume even as a tap following best practices would be nice to do if possible.

**Comment by nandahkrishna** [comment]
> The [official `wren-cli` package](https://formulae.brew.sh/formula/wren-cli) does this exact same thing.

I noticed that. It would be preferable to get it to use Homebrew's `libuv` and `wren`.

**Comment by carlocab** [comment]
> _The [official `wren-cli` package](https://formulae.brew.sh/formula/wren-cli) does this exact same thing._

It shouldn't. Thanks for tip; I'll try to look into fixing that.

**Comment by joshgoebel** [comment]
> We try to avoid vendored dependencies unless the package really cannot work without them.

Does Home-brew have a way to express lock step (exact version) dependencies?  I wonder how since I'm pretty sure packages can be freely upgraded at any time - so it'd be too easy to pull the correct version out of under someone.

> I'll try to look into fixing that.

Is it Homebrew's policy to do that even it it were against upstream package maintainers preferences/policies? 

Also how does this work when someone purposely ships a _vendored+modified_ library... as I mentioned `wren-console` does not build against `wren` proper as it adds custom extensions to the language to support necessary features.

**Comment by carlocab** [comment]
> Does Home-brew have a way to express lock step (exact version) dependencies?

No, there isn't.

> Is it Homebrew's policy to do that even it it were against upstream package maintainers preferences/policies?

[Yes](https://docs.brew.sh/Acceptable-Formulae#stuff-that-requires-vendored-versions-of-homebrew-formulae).

> Also how does this work when someone purposely ships a _vendored+modified_ library...

We'll probably wait for the modifications to be upstreamed unless something is sufficiently popular/useful to warrant overlooking the vendored dependencies.

**Comment by joshgoebel** [comment]
>> Is it Homebrew's policy to do that even it it were against upstream package maintainers preferences/policies?
>Yes.

> Homebrew formulae should avoid having multiple, separate, upstream projects bundled together in a single package to avoid shipping outdated/insecure versions of software that is already a formula.

I see, and I understand why that *usually* makes sense.   Is this true even when a single organization/person **owns all dependencies**?  Surely in that case the organization/person is best equipped to "avoid shipping outdated/insecure versions" of their own software?   

Now it's true, this doesn't apply to `uv` at all... but `wren` and `wren-cli` are both official Wren projects.  They used to be a single repository.  So far their releases have been lock-step and they are very closely integrated.  Compiling Wren-CLI 0.3 (release) against Wren 0.4 (release) makes no sense, they simply aren't compatible.  

Right now the newest versions available are:

- Wren 0.4
- Wren-CLI 0.3 (compiled against Wren 0.3) [no 0.4 release is available at this time]

Thoughts?

**Comment by joshgoebel** [comment]
> We can do this in a resource block.

Ok, that took some guesswork to figure out exactly how to stage it into place... how does this look:

https://github.com/joshgoebel/homebrew-wren/blob/main/Formula/wren-console.rb

(wrt to addressing the git/reproducable concerns)

**Comment by carlocab** [comment]
The new formula looks fine. If Wren-CLI only works with Wren 0.3 then we could probably do a `wren@0.3` formula.

**Comment by joshgoebel** [comment]
> If Wren-CLI only works with Wren 0.3 then we could probably do a wren@0.3 formula.

Well, the main branch has been upgraded to 0.4 (on git), but there is no release as of yet (and there are still a few blockers, at least in IMHO).  But yes, `wren@0.3` would be one way to go about it... would there also be `wren-cli@0.3`?  I'm not sure how much enthusiasm you'll get from upstream about killing their vendoring though.  

I already had the -slight- urge with my tap to make it depend on `libuv`... I totally see how that's super nice for Homebrew users but _it would be a huge setback for everyone else_ building the project from source.  Or perhaps Homebrew would just alter the package to build it the "homebrew way" - irregardless of how it's vendored upsteam?  

Very curious to see how this plays out.

**Comment by carlocab** [comment]
The Arch package for Wren already uses [Arch's `libuv`](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=wren#n10), so I don't think they'll find it too surprising.

Not sure I see how it's a huge setback for others. `libuv` is very easy to build, and you can always include build options that enable using a vendored `libuv` if desired (or if one isn't found in the system).

**Comment by joshgoebel** [comment]
> The Arch package for Wren already uses Arch's libuv, so I don't think they'll find it too surprising.

Oh, and I don't find that surprising at all. I guess I've just never really thought of Homebrew the same way I think of Linux distributions though... For some reason I thought Homebrew might be less opinionated.  :-)

> Not sure I see how it's a huge setback for others. 

Maybe huge is an overstatement but there is something to be said for just downloading a single repo and typing `make` and it all JUST compiles the first time, no dependencies.  I get the feeling that type of simplicity is quite valuable to Wren users.

> and you can always include build options that enable using a vendored libuv if desired 

Linking to an installed library vs compiling via vendored as an option might be nice.  I wonder if `premake` makes that easy to do?  I'm all for giving users choices. :)

**Pull request closed without merging**